Cryptodoc Update for ML-KEM (#239)
sehlen-bsi/botan-docs: docs/cryptodoc/src/05_09_ml_kem.rst

   +------------------------------+---------------------------------------------------------------------------------------------+----------------------------------------------------------------------+---------------------------+
   | ``sample_matrix``            | :srcref:`[src/lib/pubkey/kyber/kyber_common]/kyber_algos.cpp:380|sample_matrix`             | Samples a matrix from a secret seed                                  | 7, 13                     |
   +------------------------------+---------------------------------------------------------------------------------------------+----------------------------------------------------------------------+---------------------------+

Additionally, the ``PolynomialSampler`` class offers robust functionality for
sequentially sampling polynomials and polynomial vectors from a given seed
(Algorithm 8 of [FIPS-203]_).
This capability is essential for the processes of key generation and
encapsulation in ML-KEM. Table
:ref:`Polynomial Sampling Methods <pubkey/kyber/poly_sample>` lists the
supported methods. The sampling counter :math:`N` is managed by the sampler

> object and increments with each method call accordingly.

Reviewer: ```suggestion
object and is incremented with each method call accordingly.
```

   +------------------------------------------+---------------------------------------------------------------------------------------------------+---------------------------------------------------------------------+
   | ``sample_polynomial_vector_cbd_eta2``    | :srcref:`[src/lib/pubkey/kyber/kyber_common]/kyber_algos.h:86|sample_polynomial_vector_cbd_eta2`  | Polynomial vector sampling in :math:`\mathcal{D}_{\eta_2}(R_q)`     |
   +------------------------------------------+---------------------------------------------------------------------------------------------------+---------------------------------------------------------------------+

.. _pubkey/kyber/sym_primitives:

Symmetric Primitives
^^^^^^^^^^^^^^^^^^^^

In Botan, the symmetric primitives of ML-KEM are represented by the
``KyberSymmetricPrimitives`` class
(:srcref:`[src/lib/pubkey/kyber/kyber_common]/kyber_symmetric_primitives.h:30|Kyber_Symmetric_Primitives`).
This class provides an interface for the primitives, which are defined as

> :math:`PRF`, :math:`H`, :math:`J`, :math:`G`, and :math:`XOF` in Section 4.1 of

Reviewer 2: In the algorithms below, double backticks are used mostly, it seems

Author: I agree, double backticks are more consistent 👍

enabling the ``kyber`` or ``kyber_90s`` module, respectively.

Note that the Kyber 90s is already marked as deprecated, and both Kyber and
Kyber 90s may be removed as early as the next major release of the library. It
is not advisable to use any other variant than the ones specified in
[FIPS-203]_.

.. _pubkey/kyber/key_gen:

Key Generation
--------------

The high-level ML-KEM key generation (Algorithm 19) is implemented in
:srcref:`[src/lib/pubkey/kyber]/kyber_common/kyber.cpp:232|Kyber_PrivateKey::Kyber_PrivateKey`
within the ``Kyber_PrivateKey`` constructor. It delegates to the
internal and K-PKE key generation algorithms (Algorithms 16 and 13 of
[FIPS-203]_) implemented in
:srcref:`[src/lib/pubkey/kyber]/kyber_common/kyber_algos.cpp:321|expand_keypair`.
In combination, Botan does the following:

.. admonition:: Kyber_PrivateKey::Kyber_PrivateKey

   **Input:**

   -  ``rng``: random number generator
   -  ``mode``: ML-KEM mode (containing matrix dimension ``k``)

   **Output:**

   -  ``sk``: secret key
   -  ``pk``: public key

   **Steps:**

   1. Generate the random seed ``seed.d`` and the implicit rejection value ``seed.z`` at random using ``rng``

>    2. ``(rho, sigma) = G(seed.d || k)``

Reviewer 2: ML-DSA uses more math notation for rho and sigma etc (which I like) but this has to be a bit more harmonized. Also, see above, G is written in math mode in the "Symmetric primitives" section but here part of a text in double backticks.

Reviewer 2: I think I would personally prefer something like

:math:`(\rho, \sigma)` = ``G(seed.d || k)``

but I don't have very strong feelings and it applies to quite a few places.

Author: For the sake of consistency, I suggest that we stay with double backticks. This is more consistent with all other algorithms (besides ML-DSA). We will align ML-DSA with this syntax as well.

.. admonition:: ML_KEM_Encryptor::encapsulate

   **Input:**

   -  ``rng``: random number generator
   -  ``pk = (t, rho)``: public key

   **Output:**

   -  ``K``: shared secret key
   -  ``c``: ciphertext

   **Steps:**

   1. Generate a random message ``m`` using ``rng``
   2. ``(K, r) = G(m || H(pk))``
   3. K-PKE encrypt ``m`` using ``r`` to obtain ciphertext ``c``

      1. Sample transposed matrix ``At`` from ``rho`` using ``sample_matrix``
      2. Initialize a ``PolynomialSampler`` ``ps`` with ``r``
      3. ``y = ntt(ps.sample_polynomial_vector_cbd_eta1())``
      4. ``e1 = ps.sample_polynomial_vector_cbd_eta2()``
      5. ``e2 = ps.sample_polynomial_cbd_eta2()``
      6. ``u = inverse_ntt(At * y) + e1``
      7. ``mu = polynomial_from_message(m)`` for byte decoding and decompression
      8. ``v = inverse_ntt(t * y) + e2 + mu``
      9. Encode, compress and concatenate ``u`` and ``v`` to obtain the
         ciphertext ``c`` using ``compress_ciphertext``

   **Notes:**

   - Steps 1-3 correspond to Algorithms 20 and 17 of [FIPS-203]_ and are
     performed in :srcref:`[src/lib/pubkey/kyber]/ml_kem/ml_kem_impl.cpp:25|ML_KEM_Encryptor::encapsulate`.

>    - Steps 1.1-1.9 correspond to Algorithms 14 of [FIPS-203]_ and are performed

Reviewer: ```suggestion
   - Steps 3.1-3.9 correspond to Algorithms 14 of [FIPS-203]_ and are performed
```

They use the K-PKE encapsulation and decapsulation algorithms (Algorithm 14
and 15 of [FIPS-203]_) implemented in
:srcref:`[src/lib/pubkey/kyber]/kyber_common/kyber_keys.cpp:55|Kyber_PublicKeyInternal::indcpa_encrypt`
and
:srcref:`[src/lib/pubkey/kyber]/kyber_common/kyber_keys.cpp:84|Kyber_PrivateKeyInternal::indcpa_decrypt`.
In combination, Botan does the following:

.. admonition:: ML_KEM_Decryptor::decapsulate

   **Input:**

   -  ``c``: ciphertext
   -  ``sk = (seed.d, seed.z)``: secret key
   -  ``pk``: public key

   **Output:**

   -  ``K_prime``: shared secret key

   **Steps:**

   1. Recompute the secret key value ``s`` from ``seed.d``
   2. K-PKE decrypt ``c`` to obtain message ``m_prime``

      1. Retrieve ``u, v`` using ``decompress_ciphertext`` on ``c``
      2. Compute ``w = v - inverse_ntt(s * ntt(u))``
      3. ``m = polynomial_to_message(w)`` for compression and byte encoding

   3. ``(K_prime, r_prime) = G(m_prime || H(pk))``
   4. ``K_bar = J(seed.z || c)``
   5. K-PKE encrypt ``m`` using ``r_prime`` to obtain ciphertext ``c_prime``
   6. if ``c != c_prime`` set ``K_prime = K_bar``

   **Notes:**

>    - Steps 1,2 and 6-9 correspond to Algorithm 18 of [FIPS-203]_ and are

Reviewer: ```suggestion
   - Steps 1-6 correspond to Algorithm 18 of [FIPS-203]_ and are
```
